### PR TITLE
[#65] Replace slurp+parse-string with reader+parse-stream

### LINKED ISSUE #65: [performance] Remove slurp from read-json
`slurp` can have quite detrimental effects on performance. This example is with jsonista but since both it and cheshire use jackson I expect about the same order of magnitude of effect:
```clojure
(require '[criterium.core :as cc]
         '[jsonista.core :as json])
(def m {:a 1 :b 2 :c "3" :d ["5"]})
(def mapper (json/object-mapper {:decode-key-fn true}))
(def bs (json/write-value-as-bytes  m))
(type bs) ;; => [B
(defn bytes->str [^"[B" bs] (String. bs))
(bytes->str bs) ;; => "{\"a\":1,\"b\":2,\"c\":\"3\",\"d\":[\"5\"]}"
(cc/bench (json/read-value (slurp bs) mapper))
;;; Execution time mean : 8.653311 µs
(cc/bench (json/read-value (bytes->str bs) mapper))
;;; Execution time mean : 2.068923 µs
(cc/bench (json/read-value bs mapper))
;;; Execution time mean : 1.912441 µs
```
I don't mind submitting a PR with the fix.

### PR DESCRIPTION
**Pull request by bsless**
Closes #65 
Criterium benchmarks to measure overhead:
```clojure
(require
 '[criterium.core :as cc]
 '[ring.util.io :refer [string-input-stream]])

(let [handler (wrap-json-body identity)
      request  {:headers {"content-type" "application/json; charset=GBK"}
                :body (string-input-stream (String. (.getBytes "{\"foo\": \"你好\"}")) "GBK")}
      ]
  (cc/quick-bench
   (handler request)))

;;; stream Execution time mean : 5.752998 µs
;;; slurp Execution time mean : 10.297224 µs
```

### CODE REVIEW

File: src/ring/middleware/json.clj [hunk 1]
```diff
@@ -6,21 +6,23 @@
             [ring.core.protocols :as ring-protocols]
             [ring.util.io :as ring-io]
             [ring.util.response :refer [content-type]]
-            [ring.util.request :refer [character-encoding]]))
+            [ring.util.request :refer [character-encoding]])
+  (:import
```
Comment (weavejester) [inline src/ring/middleware/json.clj:10]: This `:import` foorm should be on one line.

File: src/ring/middleware/json.clj [hunk 2]
```diff
@@ -6,21 +6,23 @@
...
   (if-let [type (get-in request [:headers "content-type"])]
     (not (empty? (re-find #"^application/(.+\+)?json" type)))))
 
 (defn- read-json [request & [{:keys [keywords? bigdecimals?]}]]
   (if (json-request? request)
-    (if-let [body (:body request)]
-      (let [encoding (or (character-encoding request)
+    (if-let [^InputStream body (:body request)]
+      (let [^String encoding (or (character-encoding request)
                           "UTF-8")
```
Comment (weavejester) [inline src/ring/middleware/json.clj:21]: This line should be indented correctly.

File: src/ring/middleware/json.clj [hunk 3]
```diff
@@ -6,21 +6,23 @@
...
-            body-string (slurp body :encoding encoding)]
+            body-stream (java.io.InputStreamReader. body encoding)]
```
Comment (weavejester) [inline src/ring/middleware/json.clj:22]: `body-stream` should probably be `body-reader`.

**Review by weavejester — CHANGES_REQUESTED** [review]
Some minor formatting changes. Can you also change the commit name to:

```
Replace slurp with stream reader (fixes #65)

Improves overall performance.
```

File: src/ring/middleware/json.clj (hunk 3 above)
Comment (bsless) [reply src/ring/middleware/json.clj]: done

File: src/ring/middleware/json.clj (hunk 2 above)
Comment (bsless) [reply src/ring/middleware/json.clj]: done

File: src/ring/middleware/json.clj (hunk 1 above)
Comment (bsless) [reply src/ring/middleware/json.clj]: done

**Comment by bsless** [comment]
Done all requested changes

**Pull request merged**